boards/nrf52dk/src/aes_test.rs

```diff
@@ -0,0 +1,27 @@
+use capsules::test::aes::TestAes128Ctr;
+use kernel::hil::symmetric_encryption::{AES128, AES128_BLOCK_SIZE};
+use nrf5x::aes::{AesECB, AESECB};
+
+pub fn run() {
+    let t = static_init_test();
+
+    unsafe {
+        AESECB.set_client(t);
+    }
+
+    t.run();
+}
+
+fn static_init_test() -> &'static mut TestAes128Ctr<'static, AesECB<'static>> {
+    unsafe {
```

Comment:
This is minor, since it's in the test, but when using `unsafe`:

  * Use it _inside_ the function body, if the block exposes completely safe behavior: i.e., if there is no way to call the function in a way that could violate type-safety (for exmaple, by calling a function twice that allocates a value in static memory).

  * Use it to _mark_ a function `unsafe` if the function needs to do unsafe things and it doesn't necessarily expose a totally safe API.

I think these initializers in the test harnesses fall into the second category, so should be

```rust
unsafe fn static_init_test() -> &'static mut TestAes128Ctr<'static, AesECB<'static>> {
```

and

```rust
pub unsafe fn run()
```

they are then called from the board's `reset_handler`, which is also `unsafe` and we're good.

Reply:
That’s great rules but it is still a grey area with mutating static buffers and dereferencing memory mapped I/O (data races)

Is it ”your rules” or idiomatic Rust?

I think I haven’t followed these rules to 100% to be honest!